### PR TITLE
DevServices Support

### PR DESCRIPTION
**Pull request by stuartwdouglas**
DevServices allows for zero configuration databases in dev
and test mode, either starting the databases using
testcontainers, or creating them in process.

### CODE REVIEW

**Comment by cmoulliard** [comment]
> testcontainers

What is a `testcontainer`: a pod, a docker/podman process, ... ?

**Comment by geoand** [comment]
Awesome stuff! I'll try and play around with it tomorrow

**Comment by stuartwdouglas** [comment]
https://www.testcontainers.org/

I could do it directly using docker but would likely end up just reimplementing a lot of what is there

File: build-parent/pom.xml
Comment (famod) [inline build-parent/pom.xml:484]: I suppose you removed it because it pops up as a transitive dependency in many places now?
I wonder whether we can keep it by making the rule non-transitive? This would still cover the main problem that #13919 tried to address.

File: docs/src/main/asciidoc/datasource.adoc
Comment (maxandersen) [inline docs/src/main/asciidoc/datasource.adoc:54]: do i read this right that we'll automatically enable testcontainers in devmode ? 

what about users that do not have docker availlable ? or those who set these things outside quarkus ?

File: docs/src/main/asciidoc/datasource.adoc
Comment (maxandersen) [inline docs/src/main/asciidoc/datasource.adoc:465]:
```suggestion
you have a database extension that supports it and not config is provided Quarkus all automatically start a database (either
```

File: docs/src/main/asciidoc/datasource.adoc
Comment (stuartwdouglas) [reply docs/src/main/asciidoc/datasource.adoc:54]: If you manually configure a database then everything behaves the same way it currently does. This won't affect existing applications, except for the odd corner case of having database extensions installed but not actually using a database. 

If you don't have docker then you need to use h2 if you want to use DevDB, or just configure a database.

**Comment by knutwannheden** [comment]
This looks really nice and we are really looking forward to this (and we can stop developing our own extension which we created with the same purpose).

Question: AFAICT MSSQL is the only built-in database kind which doesn't have DevDB support. Will that still be includee in this PR?

**Comment by stuartwdouglas** [comment]
I am not sure about the licensing for MSSQL and DB2, technically it is easy to do, but I need to investigate how to handle the legalities around accepting the EULA.

**Comment by knutwannheden** [comment]
I see. As you have probably seen Testcontainers requires placing a `container-license-acceptance.txt` file on the classpath containing the Docker image name.

I suppose it should be easy for us to supply our own DevDB support for MSSQL (for our own Quarkus applications) in case you can't work the legalities out just yet.

**Comment by stuartwdouglas** [comment]
That approach is probably fine. I will check, but either way I will hold off until the initial PR is merged.

File: build-parent/pom.xml
Comment (stuartwdouglas) [reply build-parent/pom.xml:484]: I would rather keep it consistent. If we only installed it for JDK8 it would just be one more difference between the environments. getPackage was deprecated in 9 anyway.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/AugmentAction.java
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/AugmentAction.java:7]: Can we have some comments on what this does as it's not really clear by the name?

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java
Comment (geoand) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java:55]: How does this behave when test profiles are used?

File: docs/src/main/asciidoc/datasource.adoc
Comment (geoand) [reply docs/src/main/asciidoc/datasource.adoc:55]: How does DevDB behave when a `@QuarkusTestResource` is used that has configured a database?

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java
Comment (stuartwdouglas) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java:55]: It will shut down and recreate the database each time.

I am not sure if this is the correct behaviour or not, it is certainly the safest but may slow things down a bit.

**Review by geoand — COMMENTED** [review]
This is fantastic work!

As it mostly touches datasource related stuff, I think we need a review from @Sanne and / or @gsmet

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java
Comment (geoand) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java:55]: I think it's the expected behavior

File: docs/src/main/asciidoc/datasource.adoc
Comment (stuartwdouglas) [reply docs/src/main/asciidoc/datasource.adoc:55]: If you have configured a db it won't start. This removes the need for a lot of QuarkusTestResource stuff. I could have deleted them but we also need to test explicit configuration.

Comment (geoand) [reply docs/src/main/asciidoc/datasource.adoc:55]: Right, I wanted to make sure that this work doesn't break existing `@QuarkusTestResource`.

I agree that in the future a lot of that stuff will become irrelevant

File: docs/src/main/asciidoc/datasource.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/datasource.adoc:466]:
```suggestion
you have a database extension that supports it and no config is provided, Quarkus all automatically start a database (either
```

File: docs/src/main/asciidoc/datasource.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/datasource.adoc:470]:
```suggestion
application.properties and continue to use DevDB we recommend that you use the `%prod.` profile to define your
```

**Comment by maxandersen** [comment]
would it make sense to call this devcontainers or devservices instead of devdb ? or we think this approach is db specific enough and devservices for more general usecases ?

**Comment by gastaldi** [comment]
Or we could say we support testcontainers instead of coming with a new term?

**Comment by geoand** [comment]
> Or we could say we support testcontainers during dev mode instead of coming with a new term?

We do support testcontainers even without this work, so I would argue that needs an actual name like `devcontainers` that Max proposed

**Comment by gastaldi** [comment]
I see, `devcontainers` makes sense in this case if we're expecting other services to also follow (eg. Kafka, ActiveMQ servers, LDAP or SMTP servers, etc)

**Comment by geoand** [comment]
Yeah, exactly

**Comment by Sanne** [comment]
I see I was assigned as reviewer.. sorry I won't be able to get to this for at least two weeks, so please don't wait for my review that would hold progress for no good reason.

Many thanks @stuartwdouglas , at high level it looks amazing

**Comment by stuartwdouglas** [comment]
DevServices makes more sense than DevContainers, as H2 and Derby do not use containers. The fact it uses Testcontainers under the hood is more of an implementation detail, there is a chance this may eventually change.

**Comment by maxandersen** [comment]
Agreed. Devservices +1

**Comment by geoand** [comment]
I personally think that services is too overloaded, but the term seems to be used in the Kubernetes world as well, so it's probably a good choice

**Comment by stuartwdouglas** [comment]
Rename is done

**Comment by gastaldi** [comment]
This is awesome, I'll give it a try tomorrow morning. Great job @stuartwdouglas

**Comment by stuartwdouglas** [comment]
One issue I just discovered that I need to figure out a fix for is that logging does not start until RUNTIME, so log output won't appear until the container is started. If you need to pull an image this may take a while were startup has appeared to hang. I need to figure out some way to enable console logging in this case.

**Comment by stuartwdouglas** [comment]
I have pushed a potential fix to the logging problem.

**Comment by geoand** [comment]
CI seems to be complaining now:

```
2021-02-18T06:37:24.0907344Z [ERROR] Failures: 
2021-02-18T06:37:24.0909959Z [ERROR]   ConfigPropertyInjectionValidationTest Build failed with wrong exception, expected class javax.enterprise.inject.spi.DeploymentException but got java.lang.NoSuchFieldError: DELAYED_HANDLER ==> expected: <true> but was: <false>
2021-02-18T06:37:24.0913373Z [ERROR]   ClassWithMissingConfigPropertiesTest.lambda$static$1:25 expected: <javax.enterprise.inject.spi.DeploymentException> but was: <java.lang.NoSuchFieldError>
2021-02-18T06:37:24.0916769Z [ERROR]   ClassWithMissingNestedConfigPropertiesTest.lambda$static$1:25 expected: <javax.enterprise.inject.spi.DeploymentException> but was: <java.lang.NoSuchFieldError>
2021-02-18T06:37:24.0921075Z [ERROR]   InterfaceNoGetterWithMissingConfigPropertiesTest.lambda$static$1:26 expected: <javax.enterprise.inject.spi.DeploymentException> but was: <java.lang.NoSuchFieldError>
2021-02-18T06:37:24.0925318Z [ERROR]   InterfaceWithMissingConfigPropertiesTest.lambda$static$1:25 expected: <javax.enterprise.inject.spi.DeploymentException> but was: <java.lang.NoSuchFieldError>
```

**Comment by geoand** [comment]
IMHO, we should try and get this in soon to give people ample time to play with it and fine tune it before 1.13

**Comment by stuartwdouglas** [comment]
Hopefully the log CI issues are fixed now. At the moment the container logs display before the banner, but I can fix that if needed.

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:86]:
```suggestion
quarkus.datasource.db-kind=postgresql <1>
```
(not sure if Asciidoc handles things well if there's no space)

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:75]:
```suggestion
optional. If there is only one possible extension we assume this is the correct one, and if a driver has been added with test scope then
```

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:406]:
```suggestion
NOTE: Even when only one database extension is installed, named databases need to specify at least one build time
```

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:407]: Yeah, typically, that's one of the reasons why I didn't like having `db-kind` being optional. I'm still not convinced it's a good idea. I suppose we will see...

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:466]:
```suggestion
you have a database extension that supports it and no config is provided, Quarkus will automatically start a database (either
```

File: docs/src/main/asciidoc/datasource.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/datasource.adoc:463]: Should it be `DevServices`?

**Review by gsmet — CHANGES_REQUESTED** [review]
I reviewed the doc a bit and suggested a few improvements.

File: docs/src/main/asciidoc/datasource.adoc
Comment (stuartwdouglas) [reply docs/src/main/asciidoc/datasource.adoc:407]: The reason why I think it is ok is that the single DB + single driver is the 99% use case, and we can effectively make this configuration free.

**Comment by stuartwdouglas** [comment]
Doc fixes applied.

**Comment by stuartwdouglas** [comment]
What else is required to get this in?

File: bom/application/pom.xml
Comment (gastaldi) [inline bom/application/pom.xml:212]:
```suggestion
        <testcontainers.version>1.15.2</testcontainers.version>
```

Comment (gastaldi) [reply bom/application/pom.xml:212]: This is to comply with https://github.com/quarkusio/quarkus/blob/master/bom/test/pom.xml#L20

**Comment by geoand** [comment]
This is the highlight of the next release and the more baking time we have the better.
I propose we get it ASAP if there are no objections

File: extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DevServicesBuildTimeConfig.java
Comment (machi1990) [inline extensions/datasource/runtime/src/main/java/io/quarkus/datasource/runtime/DevServicesBuildTimeConfig.java:32]:
```suggestion
     * specific.
```

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I applied the latest suggestions, rebased and squashed. Let's wait for CI and merge.

**Pull request merged**
